### PR TITLE
Create usePoller hook

### PR DESCRIPTION
**Pull request by hswolff**
## ✍️ Proposed changes

🎟 _Jira ticket:_ [Name of ticket](https://jira.mongodb.org/browse/[name-of-ticket])

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] I have run `yarn changeset` and documented my changes

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README

### CODE REVIEW

File: packages/hooks/src/usePoller.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      savedCallback.current?.();
+    }
+
+    if (immediate) {
+      poll();
+    }
+
+    const id = setInterval(poll, interval);
+
+    return () => clearInterval(id);
```
Comment (jetpacmonkey) [inline packages/hooks/src/usePoller.ts]: Do we want to have any special handling for async poll functions? The mms poller will avoid firing additional requests if previous ones are still pending.

Comment (DesignerDave) [reply packages/hooks/src/usePoller.ts]: +1 – I like the idea of doing long-polling if the callback returns a promise.

File: packages/hooks/src/usePoller.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import { useState, useRef, useEffect } from 'react';
+import useEventListener from './useEventListener';
+
+function useVisibilityChange() {
+  const isVisibilityStateVisible = () => document.visibilityState === 'visible';
```
Comment (DesignerDave) [inline packages/hooks/src/usePoller.ts]: I dig this!

If we're doing this sort of fancy thing, what are your thoughts on slowing down polling on slower internet connections using the `NetworkInformation` API? (as an option for the hook, not a default of course)

Comment (hswolff) [reply packages/hooks/src/usePoller.ts]: How much should we slow polling? I assume most consumers would set their own interval, so if we exposed a hook that returned the `NetworkInformation` data then they could set the interval depending on the network speed.

Comment (DesignerDave) [reply packages/hooks/src/usePoller.ts]: Hmm... Yeah, I suppose exposing a hook for that might be a better path. I was thinking something like a `networkSpeedFactor` option that slows the polling by the passed multiple that could be less effort, though that would limit how granular a consumer could be.

Comment (jetpacmonkey) [reply packages/hooks/src/usePoller.ts]: There's no guarantee that a poller is _necessarily_ making network requests, right? Probably best to allow the consumer to make that choice.

File: packages/hooks/src/usePoller.ts (hunk 1 above)
Comment (hswolff) [reply packages/hooks/src/usePoller.ts]: Adding it in.

File: packages/hooks/src/usePoller.ts (hunk 2 above)
Comment (DesignerDave) [reply packages/hooks/src/usePoller.ts]: Yeah, I 100% agree that if we were to even bake it into usePoller, it shouldn't be enabled by default.

**Comment by hswolff** [comment]
One behavior I can't figure out what to about is...what we should do if the onPoll handler changes.

Right now we're just updating the useRef, such that the next interval will call with the new onPoll handler.

However is it fair that a person would assume that if the onPoll handler changes it should essentially 'restart' the poller and re-trigger it at the point the onPoll handler changes?

If we do that then we need to assume onPoll is memoized, which is something I don't think we can assume is true.

I'm inclined to keep the behavior as is for now. If someone needs to retrigger the onPoll themselves they can extract it to its own function and call it themselves. 

But...food for thought.

File: package.json [hunk 3]
```diff
@@ -99,7 +99,7 @@
     "file-loader": "^4.2.0",
     "gh-pages": "^2.1.1",
     "husky": "^3.0.9",
-    "jest": "23.6.0",
+    "jest": "^25.2.0",
```
Comment (hswolff) [inline package.json]: I thought upgrading jest would help with the timing tests but it didn't. But also I upgraded jest and the failure consoles are soooo much better. So I kept it.

File: scripts/jestSetup.js [hunk 4]
```diff
@@ -1,3 +1,8 @@
+// As of Jest 24 (https://jestjs.io/blog/2019/01/25/jest-24-refreshing-polished-typescript-friendly#breaking-changes)
+// this is no longer included by default for async tests
+// So we're including it so we can use async/await
+require('regenerator-runtime/runtime');
```
Comment (DesignerDave) [inline scripts/jestSetup.js]: Should we also be polyfilling this in our build? 🤔

File: packages/hooks/README.md [hunk 5]
```diff
@@ -99,3 +99,28 @@ Hook to subscribe to changes in viewport size
 ```js
 const viewportSize = useViewportSize();
 ```
+
+## usePoller
+
+Hook to create a Poller that polls at a given interval.
+
+If your `onPoll` handler returns a `Promise` it will wait for the Promise to resolve or reject before scheduling the next interval.
```
Comment (DesignerDave) [inline packages/hooks/README.md]: Is it worth noting the behavior when the window isn't active?

File: .changeset/silent-bananas-speak.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+'@leafygreen-ui/hooks': minor
+'@leafygreen-ui/text-input': patch
```
Comment (DesignerDave) [inline .changeset/silent-bananas-speak.md]: We should probably have two changesets here given the changes are unrelated.

I believe this currently will generate a changelog for text input that includes the creation of `usePoller`.

File: scripts/jestSetup.js (hunk 4 above)
Comment (hswolff) [reply scripts/jestSetup.js]: We're not using async/await in our code anymore. Probably should have some eslint rule to forbid it so we don't have to worry about it.

If we do transpile it down though I do think it should be an assumption of the application to provide their own, and we just make note of that.

File: packages/hooks/README.md (hunk 5 above)
Comment (hswolff) [reply packages/hooks/README.md]: Yeah wasn't sure where to put that. Here works.

File: .changeset/silent-bananas-speak.md (hunk 6 above)
Comment (hswolff) [reply .changeset/silent-bananas-speak.md]: Oooh that's how it works? Interesting.

File: scripts/jestSetup.js (hunk 4 above)
Comment (DesignerDave) [reply scripts/jestSetup.js]: Ahh wait, this is just for writing async functions in tests. Disregard 😅
